### PR TITLE
feat(tab): Get tabs by their ID

### PR DESCRIPTION
**Pull request by patrickrodee**
Use unique identifiers to reference tabs instead of the object equality comparison. Change the `getIndexOfTab` method to `getIndexOfTabByID`. Update tests.

This makes it more feasible to wrap MDC Tab Bar for frameworks, which shouldn't need to reference the vanilla component.

BREAKING CHANGE: `MDCTabBar#getIndexOfTab(tab: MDCTab): boolean` is now `MDCTabBar#getIndexOfTabByID(id: string): boolean`

### CODE REVIEW

File: packages/mdc-tab-bar/adapter.js [hunk 1]
```diff
@@ -25,7 +25,7 @@
 
 /* eslint-disable no-unused-vars */
 import {MDCTabDimensions} from '@material/tab/adapter';
-import {MDCTab} from '@material/tab/index';
+import {MDCTabFoundation} from '@material/tab/index';
```
Comment (kfranqueiro) [inline packages/mdc-tab-bar/adapter.js]: I would assume this is truly unused, even in annotations?

File: packages/mdc-tab-bar/adapter.js [hunk 2]
```diff
@@ -134,10 +134,10 @@ class MDCTabBarAdapter {
 
   /**
    * Returns the index of the given tab
-   * @param {!MDCTab} tab The tab whose index to determin
+   * @param {string} id The ID of the tab whose index to determine
    * @return {number}
    */
-  getIndexOfTab(tab) {}
+  getIndexOfTabByID(id) {}
```
Comment (kfranqueiro) [inline packages/mdc-tab-bar/adapter.js]: Lowercase the D (`getIndexOfTabById`)

File: packages/mdc-tab-bar/index.js [hunk 3]
```diff
@@ -170,9 +174,40 @@ class MDCTabBar extends MDCComponent {
...
+    });
+  }
+
+  /**
+   * Instantiates tab scroller component on the child tab scroller element
+   * @param {(function(!Element): !MDCTabScroller)} tabScrollerFactory
+   * @return {!MDCTabScroller=}
+   * @private
+   */
+  instantiateTabsScroller_(tabScrollerFactory) {
```
Comment (kfranqueiro) [inline packages/mdc-tab-bar/index.js]: Remove s on Tabs since the component is tab-scroller?

File: test/unit/mdc-tab-bar/mdc-tab-bar.test.js [hunk 4]
```diff
@@ -206,10 +208,10 @@ test('#adapter.getPreviousActiveTabIndex returns the index of the active tab', (
   assert.strictEqual(component.getDefaultFoundation().adapter_.getPreviousActiveTabIndex(), 1);
 });
 
-test('#adapter.getIndexOfTab returns the index of the given tab', () => {
+test('#adapter.getIndexOfTabByID; returns the index of the given tab', () => {
```
Comment (kfranqueiro) [inline test/unit/mdc-tab-bar/mdc-tab-bar.test.js]: Remove ;

File: packages/mdc-tab/index.js [hunk 5]
```diff
@@ -63,6 +66,7 @@ class MDCTab extends MDCComponent {
   initialize(
     rippleFactory = (el, foundation) => new MDCRipple(el, foundation),
     tabIndicatorFactory = (el) => new MDCTabIndicator(el)) {
+    this.id = this.root_.id;
```
Comment (kfranqueiro) [inline packages/mdc-tab/index.js]: Is the element's id guaranteed to be set by the time this runs?

File: packages/mdc-tab-bar/index.js [hunk 6]
```diff
@@ -170,9 +174,40 @@ class MDCTabBar extends MDCComponent {
...
+    return this.getTabElements_().map((el) => {
+      el.id = el.id || `mdc-tab-${++tabIdCounter}`;
+      return tabFactory(el);
+    });
+  }
+
+  /**
+   * Instantiates tab scroller component on the child tab scroller element
+   * @param {(function(!Element): !MDCTabScroller)} tabScrollerFactory
+   * @return {!MDCTabScroller=}
```
Comment (kfranqueiro) [inline packages/mdc-tab-bar/index.js]: Is this going to be problematic when the instance variable's type is `?MDCTabScroller`?

File: packages/mdc-tab-bar/index.js [hunk 7]
```diff
@@ -84,13 +86,8 @@ class MDCTabBar extends MDCComponent {
...
     this.tabFactory_ = tabFactory;
     this.tabScrollerFactory_ = tabScrollerFactory;
-
-    this.tabList_ = this.getTabElements_().map((el) => this.tabFactory_(el));
-
-    const tabScrollerElement = this.root_.querySelector(MDCTabBarFoundation.strings.TAB_SCROLLER_SELECTOR);
-    if (tabScrollerElement) {
-      this.tabScroller_ = this.tabScrollerFactory_(tabScrollerElement);
-    }
+    this.tabList_ = this.instantiateTabs_(this.tabFactory_);
```
Comment (kfranqueiro) [inline packages/mdc-tab-bar/index.js]: Can we just pass `tabFactory`? Does `this.tabFactory_` even need to exist? Ditto for `this.tabScrollerFactory_`.

File: packages/mdc-tab-bar/adapter.js (hunk 1 above)
Comment (patrickrodee) [reply packages/mdc-tab-bar/adapter.js]: Good catch

File: packages/mdc-tab/index.js (hunk 5 above)
Comment (patrickrodee) [reply packages/mdc-tab/index.js]: It is guaranteed to be set. I based this off chips which is working appropriately.

https://github.com/material-components/material-components-web/blob/6739ea0e4e7aa2cd0c93b829bade501a05eb3224/packages/mdc-chips/chip/index.js#L71

File: packages/mdc-tab-bar/index.js (hunk 6 above)
Comment (patrickrodee) [reply packages/mdc-tab-bar/index.js]: Are you referring to the typing?

File: packages/mdc-tab-bar/index.js (hunk 7 above)
Comment (patrickrodee) [reply packages/mdc-tab-bar/index.js]: Doesn't seem to be necessary in the base case.

File: packages/mdc-tab-bar/index.js (hunk 6 above)
Comment (patrickrodee) [reply packages/mdc-tab-bar/index.js]: Updated

**Pull request merged**